src/types/bytes.rs

```diff
@@ -11,6 +11,29 @@ use std::str;
 /// Represents a Python `bytes` object.
 ///
 /// This type is immutable.
+///
+/// # Equality
+///
+/// For convenience, [`Bound<'py, PyBytes>`] implements [`PartialEq<[u8]>`] to allow comparing the
+/// data in the Python bytes to a Rust `[u8]`.
+///
+/// This is not always the most appropriate way to compare Python bytes, as Python bytes subclasses
+/// may have different equality semantics. In situations where subclasses overriding equality might be
+/// relevant, use [`PyAnyMethods::eq`], at cost of the additional overhead of a Python method call.
+///
+/// ```rust
+/// # use pyo3::prelude::*;
+/// use pyo3::types::PyBytes;
+///
+/// # Python::with_gil(|py| {
+/// let py_bytes = PyBytes::new_bound(py, b"foo".as_slice());
+/// // via PartialEq<[u8]>
+/// assert_eq!(py_bytes, b"foo".as_slice());
+///
+/// // via Python equality
+/// assert!(py_bytes.as_any().eq(b"foo".as_slice()).unwrap());
```

Comment:
The problem here is, that the trait bounds turn the slice argument to `eq` into a `PyList` and not into `PyBytes`. This is a  current limitation of the generic system - we can either provide a conversion for all `&[T]` for which only `PyList` makes sense, or only a conversion for `&[u8]` which could turn into a `PyBytes`. 

So this is equivalent of the Python expression `bytes([0, 1, 2])] == [0, 1, 2]`, which also compares `false`. To fix this I think we have to create another `Bound<'_, PyBytes>` manually. It might we worth documenting this pitfall here as well.

Reply:
I think this is related to #4182, hopefully we can fix this in 0.23? 🤔

Reply:
With this in mind, what is the fix, if any, here to pass the build?

Reply:
Maybe something like this, until we can resolve that issue?
```suggestion
/// let other = PyBytes::new_bound(py, b"foo".as_slice());
/// assert!(py_bytes.as_any().eq(other).unwrap());
///
/// // Note that `eq` will convert it's argument to Python using `ToPyObject`,
/// // so the following does not compare equal since the slice will convert into a
/// // `list`, not a `bytes` object.
/// assert!(!py_bytes.as_any().eq(b"foo".as_slice()).unwrap());
```